lib/Dialect/FIRRTL/Transforms/CheckCombLoops.cpp

```diff
@@ -34,420 +25,546 @@
+                  rwProbeClasses.findLeader(srcNode))
+            continue;
+          // Check if sinkPort is a reaching def of sourcePort.
+          auto reachingDefsToSrcPort =
+              modulePaths->second.find(modSrcPortField);
+          if (reachingDefsToSrcPort != modulePaths->second.end())
+            if (llvm::find(reachingDefsToSrcPort->second, modSinkPortField) !=
+                reachingDefsToSrcPort->second.end()) {
+              // This case can occur when there are multiple RWProbes on the
+              // port, which refer to the same base value. So, each of such
+              // probes are reaching defs of each other. Hence the false
+              // loops. Instead of recording this in the reachingDefGraph,
+              // record it separately with the rwProbeClasses.
+              setOfEquivalentRWProbes.insert(srcNode);
+              if (minArgNum > srcArgNum) {
+                // Make one of the RWProbe port the base node. Use the first
+                // port for deterministic error messages.
+                minArgNum = srcArgNum;
+                basePortNode = srcNode;
+              }
+              continue;
+            }
         }
-        FieldRef instanceOutPort(inst.getResult(outPortNum),
-                                 modOutPort.getFieldID());
-        llvm::append_range(children, fieldToVals[instanceOutPort]);
+        addReachingDef(sinkPort, srcPort);
       }
-    } else if (auto mem = dyn_cast<MemOp>(ref.getDefiningOp())) {
-      if (mem.getReadLatency() > 0)
+      if (setOfEquivalentRWProbes.empty())
+        continue;
+
+      // Add all the rwprobes to the same class.
+      for (auto probe : setOfEquivalentRWProbes)
+        rwProbeClasses.unionSets(probe, sinkNode);
+
+      // Make the first port as the base value.
+      // Note: this is a port and the actual reference base exists in another
+      // module.
+      auto leader = rwProbeClasses.getLeaderValue(sinkNode);
+      rwProbeRefersTo[leader] = basePortNode;
+
+      setOfEquivalentRWProbes.insert(sinkNode);
+      // Add a reaching def  from base RWProbe port to all other RWProbe ports.
+      for (auto probe : setOfEquivalentRWProbes)
+        if (probe != basePortNode)
+          reachingDefGraph[probe].second.push_back(basePortNode);
+    }
+  }
+
+  // Record the FieldRef, corresponding to the result of the sub op
+  // `result = base[index]`
+  void recordValueRefersToFieldRef(Value base, unsigned index, Value result) {
+    auto add = [&](Value v, unsigned i) {
+      valToFieldRefs[result].emplace_back(v, i);
+    };
+
+    // Check if base is itself field of an aggregate.
+    auto it = valToFieldRefs.find(base);
+    if (it != valToFieldRefs.end()) {
+      // Rebase it to the original aggregate.
+      // Because of subaccess op, each value can refer to multiple FieldRefs.
+      for (auto &sub : it->second)
+        add(sub.getValue(), sub.getFieldID() + index);
```

Comment:
calling `add` in a loop here seems kind of slow, maybe:
```
auto &entry = valToFieldRefs[result];
for (auto &sub : it->second)
  entry.emplace_back(sub.getValue(), sub.getFieldID() + index);
```

Reply:
Thanks for the suggestion, apparently this causes to invalidate the iterator, `it` so had to introduce a temp vector and delay the update.